packages/h5wasm/src/utils.ts

```diff
@@ -19,73 +20,66 @@ export function isHDF5(buffer: ArrayBuffer): boolean {
   );
 }
 
-export function isH5WasmGroup(entity: H5WasmEntity): entity is H5WasmGroup {
-  return entity instanceof H5WasmGroup;
-}
-
-export function isH5WasmDataset(entity: H5WasmEntity): entity is H5WasmDataset {
-  return entity instanceof H5WasmDataset;
-}
-
-export function isH5WasmSoftLink(
-  entity: H5WasmEntity
-): entity is H5WasmSoftLink {
-  return entity instanceof H5WasmSoftLink;
-}
+export function convertMetadataToDType(metadata: NumericMetadata): NumericType;
+export function convertMetadataToDType(metadata: Metadata): DType;
+export function convertMetadataToDType(metadata: Metadata): DType {
+  if (isIntegerMetadata(metadata)) {
+    const { signed, size: length, littleEndian } = metadata;
+
+    return {
+      class: signed ? DTypeClass.Integer : DTypeClass.Unsigned,
+      size: length * 8,
+      endianness: littleEndian ? Endianness.LE : Endianness.BE,
+    };
+  }
 
-export function isH5WasmExternalLink(
-  entity: H5WasmEntity
-): entity is H5WasmExternalLink {
-  return entity instanceof H5WasmExternalLink;
-}
+  if (isFloatMetadata(metadata)) {
+    const { size: length, littleEndian } = metadata;
 
-export function assertH5WasmDataset(
-  entity: H5WasmEntity
-): asserts entity is H5WasmDataset {
-  if (!isH5WasmDataset(entity)) {
-    throw new Error('Expected H5Wasm entity to be dataset');
+    return {
+      class: DTypeClass.Float,
+      size: length * 8,
+      endianness: littleEndian ? Endianness.LE : Endianness.BE,
+    };
   }
-}
 
-export function assertH5WasmEntityWithAttrs(
-  entity: H5WasmEntity
-): asserts entity is H5WasmGroup | H5WasmDataset {
-  if (!isH5WasmGroup(entity) && !isH5WasmDataset(entity)) {
-    throw new Error('Expected H5Wasm entity with attributes');
-  }
-}
+  if (isStringMetadata(metadata)) {
+    const { size: length, cset } = metadata;
 
-export function convertMetadataToDType(metadata: Metadata): DType {
-  const { signed, type, size: length, littleEndian } = metadata;
+    return {
+      class: DTypeClass.String,
+      length,
+      charSet: cset === 1 ? 'UTF-8' : 'ASCII',
+    };
+  }
 
-  // See H5T_class_t in https://github.com/usnistgov/h5wasm/blob/main/src/hdf5_util_helpers.d.ts
-  switch (type) {
-    case 0:
-      return {
-        class: signed ? DTypeClass.Integer : DTypeClass.Unsigned,
-        size: length * 8,
-        endianness: littleEndian ? Endianness.LE : Endianness.BE,
-      };
+  if (isCompoundMetadata(metadata)) {
+    const { compound_type } = metadata;
+    const { members, nmembers } = compound_type;
 
-    case 1:
-      return {
-        class: DTypeClass.Float,
-        size: length * 8,
-        endianness: littleEndian ? Endianness.LE : Endianness.BE,
-      };
+    if (nmembers === 2 && members[0].name === 'r' && members[1].name === 'i') {
+      const [realTypeMetadata, imagTypeMetadata] = members;
+      assertNumericMetadata(realTypeMetadata);
+      assertNumericMetadata(imagTypeMetadata);
```

Comment:
Thanks to this assertion and the overload of `convertMetadataToDType` above, TS knows that converting this will yield a `NumericType` removing the need to assert after conversion.